src/Components/InMemoryEventStream.cs

```diff
@@ -0,0 +1,74 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Components
+{
+    internal sealed class InMemoryEventStream : EventStream, EventDispatcher
+    {
+        private readonly HashSet<object> _subscriptions = new HashSet<object>();
+
+        public void Dispatch<T>(T entity)
+        {
+            if (entity == null)
+                throw new ArgumentNullException(nameof(entity));
+
+            lock(_subscriptions)
+                foreach (var subscription in _subscriptions.OfType<IObserver<T>>())
+                    subscription.OnNext(entity);
+        }
+
+        public IDisposable Subscribe<T>(Action<T> callback)
+        {
+            if (callback == null)
+                throw new ArgumentNullException(nameof(callback));
+
+            return Subscribe(new CallbackBasedObserver<T>(callback));
+        }
+
+        public IDisposable Subscribe<T>(IObserver<T> subscriber)
+        {
+            if (subscriber == null)
+                throw new ArgumentNullException(nameof(subscriber));
+
+            lock(_subscriptions)
+                _subscriptions.Add(subscriber);
+
+            return new Subscription(() => Unsubscribe(subscriber));
+        }
+
+        private void Unsubscribe<T>(IObserver<T> subscriber)
+        {
+            lock (_subscriptions)
+                _subscriptions.Remove(subscriber);
+        }
+
+        private class CallbackBasedObserver<T> : IObserver<T>
+        {
+            private readonly Action<T> _callback;
+
+            internal CallbackBasedObserver(Action<T> callback)
+            {
+                _callback = callback;
+            }
+            public void OnCompleted()
```

Comment:
I removed the previous lines between methods comment because I know your commits won't have done it automatically on here. However, here's another one. :)

Personally, empty methods I keep on a single line but if you don't want to. i.e. `public void OnCompleted() { }`